conda_mirror/diff_tar.py

```diff
@@ -0,0 +1,191 @@
+import os
+import sys
+import json
+import hashlib
+import tarfile
+from os.path import abspath, isdir, join, relpath
+
+
+mirror_dir = None
+reference_path = './reference.json'
+
+
+def md5_file(path):
+    """
+    Return the MD5 hashsum of the file given by `path` in hexadecimal
+    representation.
+    """
+    h = hashlib.new('md5')
+    with open(path, 'rb') as fi:
+        while 1:
+            chunk = fi.read(262144)
+            if not chunk:
+                break
+            h.update(chunk)
+    return h.hexdigest()
+
+
+def find_repos():
+    """
+    Asssuming the global `mirror_dir` is set, iterate all sub-directories
+    which contain a repodata.json and repodata.json.bz2 file.
+    """
+    for root, unused_dirs, files in os.walk(mirror_dir):
+        if 'repodata.json' in files and 'repodata.json.bz2' in files:
+            yield root
+
+
+def all_repodata():
+    """
+    Return a dictionary mapping all repository sub-directories to the conda
+    package list as respresented by the 'packages' field in repodata.json.
+    """
+    d = {}
+    for repo_path in find_repos():
+        with open(join(repo_path, 'repodata.json')) as fi:
+            index = json.load(fi)['packages']
+        d[repo_path] = index
+    return d
+
+
+def verify_all_repos():
+    """
+    Verify all the MD5 sum of all conda packages listed in all repodata.json
+    files in the repository.
+    """
+    d = all_repodata()
+    for repo_path, index in d.items():
+        for fn, info in index.items():
+            path = join(repo_path, fn)
+            if info['md5'] == md5_file(path):
+                continue
+            print('MD5 mismatch: %s' % path)
+
+
+def write_reference():
+    """
+    Write the "reference file", which is a collection of the content of all
+    repodata.json files.
+    """
+    data = json.dumps(all_repodata(), indent=2, sort_keys=True)
+    # make sure we have newline at the end
+    if not data.endswith('\n'):
+        data += '\n'
+    with open(reference_path, 'w') as fo:
+        fo.write(data)
+
+
+def read_reference():
+    """
+    Read the "reference file" from disk and return its content as a dictionary.
+    """
+    try:
+        with open(reference_path) as fi:
+            return json.load(fi)
+    except FileNotFoundError:
+        sys.exit('No such file: %s' % reference_path)
```

Comment:
You probably shouldn't call `sys.exit` from a function like this.  Just raise a system exit or similar

Reply:
Fixed